src/connection/index.js

```diff
@@ -0,0 +1,456 @@
+    })
+    this._state.on('CONNECTED', () => {
+      this.log(`unmuxed connection opened to ${this.theirB58Id}`)
+      this.emit('unmuxed', this.conn)
+    })
+    this._state.on('DISCONNECTING', () => this._onDisconnecting())
+    this._state.on('ABORTED', () => this._onAborted())
+    this._state.on('ERRORED', () => this._onErrored())
+    this._state.on('error', (err) => this._onStateError(err))
+  }
+
+  /**
+   * Gets the current state of the connection
+   *
+   * @returns {string} The current state of the connection
+   */
+  getState () {
+    return this._state._state
+  }
+
+  /**
+   * Puts the state into dialing mode
+   *
+   * @fires ConnectionFSM#Error May emit a DIAL_SELF error
+   * @returns {void}
+   */
+  dial () {
+    if (this.theirB58Id === this.ourPeerInfo.id.toB58String()) {
+      return this.emit('error', Errors.DIAL_SELF())
+    }
+
+    this._state('dial')
+  }
+
+  /**
+   * Initiates a handshake for the given protocol
+   *
+   * @param {string} protocol The protocol to negotiate
+   * @param {function(Error, Connection)} callback
+   * @returns {void}
+   */
+  shake (protocol, callback) {
+    // If there is no protocol set yet, don't perform the handshake
+    if (!protocol) {
+      return callback(null, null)
+    }
+
+    if (this.muxer && this.muxer.newStream) {
+      return this.muxer.newStream((err, stream) => {
+        if (err) {
+          return callback(err, null)
+        }
+
+        this.log(`created new stream to ${this.theirB58Id}`)
+        this._protocolHandshake(protocol, stream, callback)
+      })
+    }
+
+    this.conn.setPeerInfo(this.theirPeerInfo)
+    this._protocolHandshake(protocol, this.conn, callback)
+  }
+
+  /**
+   * Puts the state into muxing mode
+   *
+   * @returns {void}
+   */
+  upgrade () {
+    this._state('upgrade')
+  }
+
+  /**
+   * Event handler for dialing. Transitions state when successful.
+   *
+   * @private
+   * @fires ConnectionFSM#error
+   * @returns {void}
+   */
+  _onDialing () {
+    // TODO: Start the connection flow
+    // TODO: Allow multiple dials?
+    this.log(`dialing ${this.theirB58Id}`)
+
+    if (!this.switch.hasTransports()) {
+      this.emit('error', Errors.NO_TRANSPORTS_REGISTERED())
+      return this._state('disconnect')
+    }
+
+    const tKeys = this.switch.availableTransports(this.theirPeerInfo)
+
+    const circuitEnabled = Boolean(this.switch.transports[Circuit.tag])
+    let circuitTried = false
+
+    const nextTransport = (key) => {
+      let transport = key
+      if (!transport) {
+        if (!circuitEnabled) {
+          this.emit('error', new Error(
+            `Circuit not enabled and all transports failed to dial peer ${this.theirB58Id}!`
+          ))
+          return this._state('disconnect')
+        }
+
+        if (circuitTried) {
+          this.emit('error', new Error(`No available transports to dial peer ${this.theirB58Id}!`))
+          return this._state('disconnect')
+        }
```

Comment:
I like the warning option as that would provide the user with more information about the various attempts, even if connections succeed on a later transport.

With the current structure the switch would need to emit the events. Perhaps the payload should also include the peerInfo so users understand where it came from. Something like:
```js
_switch.emit('error:connection_attempt_failed', {
  peerInfo: theirPeerInfo,
  error: ...
})
```

Reply:
I see. This dialing is triggered by a `switch.dial()`, which is callback-based..
Having the switch emit these events independently of the dial() call means that, for the API user,  it will be hard to correlate a call with these warnings.
Usually I solve this by either returning an even stream or an event emitter. In this case, I think that something like a dialFSM would be something useful to consider...

Reply:
That's a good point. Currently `Libp2p` doesn't return anything from its dial methods so returning an event emitter there from a `Switch` dialFSM would enable us to keep backwards compatibility more easily while we transition the system.

For js-ipfs we'd also need to make sure to return that emitter from [`swarm.connect()`](https://github.com/ipfs/js-ipfs/blob/c73bd2f68a57651e2b7a67b522f79556040963ae/src/core/components/swarm.js#L71).

Reply:
@pgte I've added a `dialsFSM` method to the switch, that can be used instead of `dial`. Instead of a connection, it will return the fsm. I've added some tests for it to show the events and functionality it has https://github.com/libp2p/js-libp2p-switch/pull/278/files#diff-080e21482eea9ffdd7ee511eb71d3522.

**The main things to note**:
The raw connection is still returned in the callback passed to dialFSM
1. Users can listen on `error` events to get fatal messages
2. Users can listen on `error:connection_attempt_failed` events to get a list of errors for failed transport dials, these are not fatal to the connection.
3. Users can listen on `close` events to know when the specific connection is closed.
4. Users can call `fsm.close()` to close that specific connection.

Are there other events or methods that would be helpful to have? We can always wait to add other things if there aren't higher priority nice to haves.

Reply:
@jacobheun That's great! Other than successful connection attempt (which you can infer from having the "connected" callback called), I don't see any other so far.